File: unisync_android/app/src/main/java/com/anhquan/unisync/core/plugins/volume/VolumePlugin.kt
package com.anhquan.unisync.core.plugins.volume

import com.anhquan.unisync.core.Device
import com.anhquan.unisync.core.DeviceConnection
import com.anhquan.unisync.core.plugins.UnisyncPlugin
import com.anhquan.unisync.models.DeviceMessage

class VolumePlugin(
    private val device: Device,
) : UnisyncPlugin(device, DeviceMessage.Type.VOLUME) {
    private object Method {
        const val GET_VOLUME = "get_volume"
        const val SET_VOLUME = "set_volume"
        const val VOLUME_CHANGED = "volume_changed"
    }

    init {
        sendRequest(Method.GET_VOLUME)
    }

    override fun listen(
        header: DeviceMessage.DeviceMessageHeader,
        data: Map<String, Any?>,
        payload: DeviceConnection.Payload?
    ) {
        if (header.method == Method.VOLUME_CHANGED) {
            notifier.onNext(data)
        }
    }

    fun changeVolume(value: Float) {
        sendRequest(
            Method.SET_VOLUME,
            mapOf(
                "value" to value
            )
        )
    }
}

File: lib/core/plugins/volume/volume.plugin.dart
import 'dart:async';

import 'package:flutter_volume_controller/flutter_volume_controller.dart';
import 'package:unisync/core/device.dart';
import 'package:unisync/core/device_connection.dart';
import 'package:unisync/core/plugins/base_plugin.dart';
import 'package:unisync/models/device_message/device_message.model.dart';

class VolumePlugin extends UnisyncPlugin {
  VolumePlugin(Device device) : super(device, type: DeviceMessage.Type.VOLUME) {
    _subscription = FlutterVolumeController.addListener((value) {
      if (!_isChangingByPeer) {
        sendNotification(
          _Method.VOLUME_CHANGED,
          data: {'volume': value},
        );
      }
    });
  }

  static const _Method = (
    GET_VOLUME: 'get_volume',
    SET_VOLUME: 'set_volume',
    VOLUME_CHANGED: 'volume_changed'
  );

  late final StreamSubscription<double> _subscription;

  bool _isChangingByPeer = false;
  late final debounce = _Debouncer(
    // duration: const Duration(seconds: 2),
    callback: () {
      _isChangingByPeer = false;
    },
  );

  @override
  Future<void> onReceive(
    DeviceMessageHeader header,
    Map<String, dynamic> data,
    Payload? payload,
  ) async {
    super.onReceive(header, data, payload);
    if (header.method == _Method.SET_VOLUME) {
      _setVolume(data['set_volume']);
    }
    if (header.method == _Method.GET_VOLUME) {
      sendNotification(
        _Method.VOLUME_CHANGED,
        data: {
          'volume': await FlutterVolumeController.getVolume(),
        },
      );
    }
  }

  Future<void> _setVolume(double value) async {
    await FlutterVolumeController.setVolume(value);
    _isChangingByPeer = true;
    debounce.execute();
  }

  @override
  void dispose() {
    _subscription.cancel();
    super.dispose();
  }
}

class _Debouncer {
  _Debouncer(
      {this.duration = const Duration(milliseconds: 500),
      required this.callback});

  final void Function() callback;
  final Duration duration;
  Timer? timer;

  void execute() {
    if (timer == null || timer!.isActive) {
      timer?.cancel();
      timer = Timer(duration, callback);
    } else {
      timer = Timer(duration, callback);
    }
  }
}
